File: 2560-house-robber-iv/2560-house-robber-iv.cpp
class Solution {
public:
    bool isPossible(vector<int> &a, int k, int req){
        vector<int> dp(a.size());
        for(int i=0;i<a.size();i++){
            dp[i]=-1;
        }
        int maxi=help(a,k,0,dp);
        return maxi>=req;
    }
    int help(vector<int> &arr, int k, int i, vector<int> &v){
        if(i>=arr.size())return 0;
        if(v[i]!=-1)return v[i];
        int temp=help(arr,k,i+1,v);
        if(arr[i]<=k){
            return v[i]= max(help(arr,k,i+2,v)+1,temp);
        }
        return v[i]= temp;
    }
    
    int minCapability(vector<int>& nums, int k) {
        int l=0, r=1e9;
        while(l<r-1){
            int mid=l+(r-l)/2;
            if(isPossible(nums,mid,k))
                r=mid;
            else l=mid;
   
        }
        return r;
    }
};

File: 2560-house-robber-iv/2560-house-robber-iv.js
var minCapability = function(nums, k) {
  let min = nums[0], max = nums[0];
  let n = nums.length;
  for (let i = 0; i < n; i++) {
    min = Math.min(min, nums[i]);
    max = Math.max(max, nums[i]);
  }
  let low = min, high = max;
  while (low < high) {
    let mid = Math.floor((low + high) / 2);
    if (isEnough(mid)) high = mid;
    else low = mid + 1;
  }
  return low;
  
  function isEnough(max) { 
    let houses = 0;
    for (let i = 0; i < n; i++) {
      if (nums[i] <= max) {
        houses++;
        i++;
      }
      if (houses === k) return true;
    }
    return false;
  }
};